improvement: S3UTILS-86 helper to filter recently updated keys

jonathan-gramain opened the pull request:
Create a helper class DiffStreamOplogFilter that tails all raft
sessions from the specified cseq offset in the background, while
ingesting diff entries. It then only forwards those which do not
appear in their respective raft session's oplog to the stream's
output.

jonathan-gramain committed:
improvement: S3UTILS-86 helper to filter recently updated keys

Create a helper class DiffStreamOplogFilter that tails all raft
sessions from the specified cseq offset in the background, while
ingesting diff entries. It then only forwards those which do not
appear in their respective raft session's oplog to the stream's
output.

watsaqat commented on CompareRaftMembers/DiffStreamOplogFilter.js:
```diff
@@ -0,0 +1,240 @@
+                bucketdHost,
+                bucketdPort,
+                raftSessionId: rsId,
+                startSeq: excludeFromCseqs[rsId] + 1,
```
a comment here about why startSeq is defined as such would be helpful.

jonathan-gramain replied:
done

watsaqat commented on CompareRaftMembers/DiffStreamOplogFilter.js:
```diff
@@ -0,0 +1,240 @@
+        this.raftSessionStates = {};
+        for (const rsId of Object.keys(excludeFromCseqs)) {
+            const raftSessionState = {};
+            this.raftSessionStates[rsId] = raftSessionState;
```
this might be better at the bottom of the block after raftSessionState is initialized with oplogStream, oplogKeys, inputBuffer

watsaqat commented on CompareRaftMembers/DiffStreamOplogFilter.js:
```diff
@@ -0,0 +1,240 @@
+                if (rsId !== undefined) {
+                    return next(null, rsId);
+                }
+                return this._fetchBucketRaftSessionId(bucketName, next);
```
I would just move the behavior of lines 97-100 inside of _fetchBucketRaftSessionId

watsaqat commented on CompareRaftMembers/DiffStreamOplogFilter.js:
```diff
@@ -0,0 +1,240 @@
+            return callback();
+        }
+        this.pendingFlushCallback = callback;
+        return undefined;
```
why the explicit undefined return?

jonathan-gramain replied:
one of our linter rules is `consistent-return`, which is making sure that if we return somewhere (even `undefined`, here at https://github.com/scality/s3utils/pull/246/files#diff-8275bd7edd92cb9b8c115839e21be4e42f8d5e708502eff30fcf85566bd6793dR137), we have to return something everywhere in the function.

jonathan-gramain replied:
We abuse the `return callback()` notation though, since we don't care about the returned value.

watsaqat commented on CompareRaftMembers/DiffStreamOplogFilter.js:
```diff
@@ -0,0 +1,240 @@
+        }
+    }
+
+    /**
```
I'd probably just leave this jsDoc out since the rest of the file doesn't use them (unless you want to add them elsewhere for consistency and describe diffEntry shape)

jonathan-gramain replied:
I think it's fine to leave this JSDoc since it can help understanding the purpose of this helper function. I actually think the `DiffStreamOplogFilter` class could benefit from a high-level JSDoc as well, to explain its concept and its input/output, which is aligned to what you're saying.

watsaqat commented on CompareRaftMembers/DiffStreamOplogFilter.js:
```diff
@@ -0,0 +1,240 @@
+                if (res.statusCode !== 200) {
+                    return cb(new Error(`GET ${reqParams.path} returned status ${res.statusCode}`));
+                }
+                return cb(null, { statusCode: res.statusCode, body });
```
statusCode should always be 200 here since we return non-200s above, right? I think just 'statusCode: 200' is easier to scan through

jonathan-gramain replied:
indeed, will change it

jonathan-gramain committed:
S3UTILS-86 address review comments

- address some review comments

- add JSDoc to describe the DiffStreamOplogFilter class

bert-e commented:
# Hello jonathan-gramain,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/s3utils/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

jonathan-gramain commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/1.13`



The following branches will **NOT** be impacted:

* `development/1.12`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/1.13`



The following branches have **NOT** changed:

* `development/1.12`



Please check the status of the associated issue S3UTILS-86.

Goodbye jonathan-gramain.
